src/sentry/integrations/gitlab/client.py

```diff
@@ -42,20 +42,32 @@ def metadata(self):
         return self.installation.model.metadata
 
     def request(self, method, path, data=None, params=None, api_preview=False):
-        # TODO(lb): Refresh auth
-        # self.check_auth(redirect_url=self.oauth_redirect_url)
         access_token = self.identity.data['access_token']
         headers = {
             'Authorization': u'Bearer {}'.format(access_token)
         }
-        return self._request(
-            method,
-            GitLabApiClientPath.build_api_url(
-                self.metadata['base_url'],
-                path
-            ),
-            headers=headers, data=data, params=params
+        url = GitLabApiClientPath.build_api_url(
+            self.metadata['base_url'],
+            path
         )
+        try:
+            return self._request(
+                method,
+                url,
+                headers=headers, data=data, params=params
+            )
+        except ApiUnauthorized:
+            self.update_auth()
+            return self._request(
+                method,
+                url,
+                headers=headers, data=data, params=params
+            )
+
+    def update_auth(self):
+        # TODO(lb): This is hard to find.
+        # The docs for whatever reason do not cover refresh!
+        pass
```

Comment:
I dig a little digging – GitLab uses a gem called Doorkeeper for auth. In Doorkeeper's wiki, they have some information about refreshing tokens:

https://github.com/doorkeeper-gem/doorkeeper/wiki/Enable-Refresh-Token-Credentials#testing-with-oauth2-gem

In the last code example in that section, they hint at what you need to do. I also poked around the tests for Doorkeeper (tests are a good place to look for how something works, since ideally, that what tests describe) and found:

https://github.com/doorkeeper-gem/doorkeeper/blob/946bbbd0f1be042f047c6e589e79a87df524649b/spec/requests/flows/refresh_token_spec.rb#L28

So it looks like it's the same `/oauth/token` endpoint we use to create one in the first place, but we set the `grant_type` param to `refresh_token` and include a `refresh_token` param with a value of the refresh token we currently have.

This will return a new access token _and_ refresh token, I believe.

Reply:
@mnoble What you described matches how I've worked with other doorkeeper based apps in the past.